cheatsheets/Nodejs_security_cheat_sheet.md

```diff
@@ -0,0 +1,371 @@
+
+Errors in these callbacks can be propagated as many times as possible. Each callback that the error has been propagated to can ignore, handle or propagate the error.
+
+## Listen to errors when using EventEmitter
+
+When using EventEmitter, errors can occur anywhere in the event chain. Normally, if an error occurs in an EventEmitter object, an error event with an Error object as its argument is called. However, if there are no attached listeners to that error event, the Error object that is sent as argument is thrown and becomes an uncaught exception. In short, if you do not handle errors within an EventEmitter object properly, these unhandled errors may crash your application. Therefore, you should always listen to error events when using EventEmitter objects.
+
+```JavaScript
+var events = require(‘events’);
+var myEventEmitter = function(){
+    events.EventEmitter.call(this);
+}
+require(‘util’).inherits(myEventEmitter, events.EventEmitter);
+myEventEmitter.prototype.someFunction = function(param1, param2) {
+    //in case of an error
+    this.emit(‘error’, err);
+}
+var emitter = new myEventEmitter();
+emitter.on(‘error’, function(err){
+    //Perform necessary error handling here
+});
+```
+
+## Handle uncaughtException
+
+Node.js behavior for uncaught exceptions is to print current stack trace and then terminate the thread. However, Node.js allows customization of this behavior. It provides a global object named process which is available to all Node.js applications. It is an EventEmitter object  and in case of an uncaught exception, “uncaughtException” event gets emitted and it is brought up to the main event loop. In order to provide a custom behavior for uncaught exceptions, you can bind to this event. However, resuming the application after such an uncaught exception can lead to further problems. Therefore, if you do not want to miss any uncaught exception, you should bind to uncaughtException event and cleanup any allocated resources like file descriptors, handles and similar before shutting down the process. Resuming the application is strongly discouraged as the application will be in an unknown state.
+
+```JavaScript
+process.on(“uncaughtException”, function(err) {
+    // clean up allocated resources
+    // log necessary error details to log files
+    process.exit(); // exit the process to avoid unknown state
+});
+```
+
+## Monitor the event loop
+
+When your application server is under heavy network traffic, it may not be able to serve its users. This is essentially a type of Denial of Service (DoS) attack. Toobusy module allows you to monitor the event loop. It keeps track of lags and when it goes beyond a certain threshold, this module can indicate your server is too busy. In that case, you can stop processing incoming requests and send them 503 Server Too Busy message so that your application stay responsive. Sample use of toobusy module is shown here:
```

Comment:
`_The_ Toobusy module`

`keeps track of the response time, and when`...

`Example` rather than `Sample`

Reply:
changed accordingly